### DISCUSSION TITLE
Need an advice on how to read the reed switch using the Engine Top Hat and/or Sailor ESP32 Hat

### DISCUSSION THREAD

**Discussion by sergei** [hatlabs/discussions#50] [Q&A] [answered]
Need an advice on what is the best way to read the wind speed of B&G Masthead unit. Apparently it has the reed switch that closes every revolution of wind cap. The B&G Wind instrument has  5V on the pin connected to this sensor and most likely senses when it's connected to the ground by the switch. I want to do the same using Sailor Hat.

I guess one way would be to use the the Analog Input in current source mode, but then I need to poll them at quite fast rate to be accurate.

I would prefer to have an interrupt generated every time the switch either opens or closes, which can be done with some GPIO and PCNT unit of ESP32. Is there any feature Engine Hat or Sailor ESP32 Hat I can use directly without any HW added?

**Comment by sergei** [comment]
Just to add some color to it. 
If I set up GPIO13 with pull up and then short the DI1 to ground that what I see in DI1:
![image](https://user-images.githubusercontent.com/138027/235485928-056d19e0-d1c0-46f7-865d-2dafee98f85b.png)
But if I check the GPIO13 itself:
![image](https://user-images.githubusercontent.com/138027/235486046-b1facb16-9374-4f86-a204-57910470afd7.png)
As a result I don't have any interrupt triggering when DI1 is connected to the ground

**Reply by mairas** [reply]
Hi Sergei,

The analog inputs have a strong low-pass filtering on them and probably wouldn't work for this application.

The series input resistor on the digital inputs is quite large (100k) and if you enable the internal pull-up on the GPIO, it's likely that the input won't be pulled down (as you noticed). Try setting the GPIO as INPUT without the pullup; I bet that should work.

**Reply by sergei** [reply]
Indeed originally I had the GPIO configured as floated (gpio_pull_mode_t == GPIO_FLOATING). It didn't work either. I don't see DI1 pulled up with that settings it floats close to ground.

**Reply by mairas** [reply, marked as answer]
Yes, there is a weak pull-down resistor that does that. So the signal is pulled down when the reed switch is closed and open circuit when it's open? Adding a suitable (1-10k, maybe) physical pull-up resistor before the engine hat input might do the trick.

**Reply by sergei** [reply]
Yes, I put 20K pull up on In1, set the gpio_pull_mode_t == GPIO_FLOATING and it seems working OK. Now need to find the graceful way to solder this resistor. That a lot!

**Reply by mairas** [reply]
Put the resistor on the proto area, one header pin at each end and connect with [wire wraps](https://en.wikipedia.org/wiki/Wire_wrap) or a piece of solder!

**Reply by sergei** [reply]
Yes, did just that:
![image](https://user-images.githubusercontent.com/138027/235727027-00f2925b-b9d3-4cac-be88-742dbe4486dc.png)
Here is fully assembled setup:
![image](https://user-images.githubusercontent.com/138027/235727272-3149c861-6238-4a00-b60c-eca26567568a.png)

Don't really trust these wire to stay on header pins while on a boat in rough seas. Any advice how to secure them?

**Reply by mairas** [reply]
The header pins are indeed not super-reliable. One approach is to remove the black plastic housing and add a bit of solder to the joint and cover everything with heat-shrink tube.

Or, you could use my favorites, wire wraps.

Get a tool like this: https://www.amazon.com/Singlestranded-Practical-Function-Repairing-WireWsu30M/dp/B08H9P1R1L
and some wire wrap wire like this: https://www.amazon.com/Colored-Insulation-B-30-1000-8-Wire-Wrapping/dp/B07BLZ333V/

And then connect any pins together with a bit of wire and the wire wrapping tool. This video gives an excellent overview:
https://www.youtube.com/watch?v=L-463vchW0o
